### ISSUE TITLE
shap for unsupervised model

### ISSUE THREAD

**Issue by cmottac** [shap/shap#1052]
Dear all,
is it possible to use shap to get explanations for unsupervised models? 
I tried to apply the various explainer to clustering algorithms including as kmeans,agglomerative clustering, (h)dbscan, Birch, but I always got errors.

Thank you in advance.

**Comment by slundberg** [comment]
SHAP is a feature attribution method, which means it attributes to a set of input features responsibility for the output of a function that depends on those features. That function does not have to be a supervised machine learning model, but it does need some kind of numeric output. Unsupervised clustering algorithms output a class label for each input data sample, so you could try and explain that label with PartitionExplainer or KernelExplainer (one hot encoded might make the most sense). An example of a directly supported unsupervised model is sklearn.ensemble.IsolationForest which is supported by TreeExplainer.

**Comment by cmottac** [comment]
Thank you @slundberg.  
I think what I was doing wrong is to pass to the KernelExplainer the clusterer object, rather than the predict function. I am now able to use it with kmeans. In include the minimal code, which may be useful for future readers:
```
import sklearn, shap, matplotlib
blobs, labels = sklearn.datasets.make_blobs(n_samples=[100,100,100], n_features=2,random_state=49)
clusterer = sklearn.cluster.KMeans(n_clusters=3).fit(blobs)
matplotlib.pyplot.scatter(blobs[:,0],blobs[:,1],s=6,c=clusterer.labels_); plt.show()
explainer = shap.KernelExplainer(model=clusterer.predict,data=blobs)
shap_values = explainer.shap_values(blobs)
shap.initjs()
shap.summary_plot(shap_values, blobs)
```
This solves my issue.
As a side question: when I try to do the same with hdbscan, I get a weird summary plot, which lead me thinking that I am doing something wrong:
```
import sklearn, hdbscan, shap, matplotlib
blobs, labels = sklearn.datasets.make_blobs(n_samples=[100,100,100], n_features=2,random_state=49)
clusterer = hdbscan.HDBSCAN(prediction_data=True).fit(blobs)
matplotlib.pyplot.scatter(blobs[:,0],blobs[:,1],s=6,c=clusterer.labels_); plt.show()
def hdbscan_predict(X):
    return hdbscan.approximate_predict(clusterer,X)[0]
explainer = shap.KernelExplainer(model=hdbscan_predict,data=blobs)
shap_values = explainer.shap_values(blobs)
shap.initjs()
shap.summary_plot(shap_values, blobs)
```
If anyone is interested in checking this out, let me know!
Thank you,
Carlo

**Comment by slundberg** [comment]
The summary plot is weird because the order of the classes for hdbscan does not line up along the feature axes like it happens to for the kmeans output. I would explain each cluster membership separately like: 
```
def hdbscan_predict(X):
    return hdbscan.approximate_predict(clusterer,X)[0] == 0
```
(which explains if a sample is in the 0-th cluster)
